File: backend/app/services/critic.py
from dataclasses import dataclass

from app.models.schemas import Source


@dataclass
class CriticResult:
    passed: bool
    reason: str


class CriticService:
    def evaluate(self, answer: str, sources: list[Source]) -> CriticResult:
        if not answer.strip():
            return CriticResult(passed=False, reason="Empty answer")

        if "api key is not configured" in answer.lower():
            return CriticResult(passed=True, reason="Fallback mode")

        if not sources:
            return CriticResult(passed=False, reason="No retrieved sources")

        answer_terms = set(self._normalize(answer))
        source_terms: set[str] = set()
        for source in sources:
            source_terms.update(self._normalize(source.snippet))
            source_terms.update(self._normalize(source.title))

        if not answer_terms:
            return CriticResult(passed=False, reason="Answer has no lexical signal")

        overlap = len(answer_terms.intersection(source_terms)) / max(1, len(answer_terms))
        if overlap < 0.08:
            return CriticResult(passed=False, reason="Low grounding overlap")

        return CriticResult(passed=True, reason="Grounded by retrieved evidence")

    def _normalize(self, text: str) -> list[str]:
        cleaned = "".join(char.lower() if char.isalnum() else " " for char in text)
        return [token for token in cleaned.split() if len(token) > 2]

File: nextapp/lib/critic.js
function normalize(text) {
  return text
    .replace(/[^a-z0-9\s]/g, ' ')
    .toLowerCase()
    .split(/\s+/)
    .filter((t) => t.length > 2);
}

/**
 * @param {string} answer
 * @param {Array<{title: string, snippet: string}>} sources
 * @returns {{ passed: boolean, reason: string }}
 */
export function evaluateCritic(answer, sources) {
  if (!answer.trim()) {
    return { passed: false, reason: 'Empty answer' };
  }

  if (answer.toLowerCase().includes('api key is not configured')) {
    return { passed: true, reason: 'Fallback mode' };
  }

  if (!sources || sources.length === 0) {
    return { passed: false, reason: 'No retrieved sources' };
  }

  const answerTerms = new Set(normalize(answer));
  const sourceTerms = new Set();
  for (const s of sources) {
    for (const t of normalize(s.snippet || '')) sourceTerms.add(t);
    for (const t of normalize(s.title || '')) sourceTerms.add(t);
  }

  if (answerTerms.size === 0) {
    return { passed: false, reason: 'Answer has no lexical signal' };
  }

  let overlap = 0;
  for (const t of answerTerms) {
    if (sourceTerms.has(t)) overlap++;
  }

  const ratio = overlap / answerTerms.size;
  if (ratio < 0.08) {
    return { passed: false, reason: 'Low grounding overlap' };
  }

  return { passed: true, reason: 'Grounded by retrieved evidence' };
}
